### PR TITLE
Feature hcc profile

### PR DESCRIPTION
**Pull request by bensander**
Simple but useful GPU command profiler for HCC.  Enabled with env var, shows kernel and data  xfer times and bandwidth.

### CODE REVIEW

**Comment by aaronenyeshi** [comment]
Hi, is there documentation on these debuggers and profilers? I think it would be very useful to add how to use this tool (such as which env var to set) into debugging documentations.

File: lib/hsa/mcwamp_hsa.cpp [hunk 1]
```diff
@@ -41,6 +41,7 @@
 #include <time.h>
 #include <iomanip>
 
+#define KALMAR_DEBUG (0)
```
Comment (whchung) [inline lib/hsa/mcwamp_hsa.cpp]: @bensander please help remove this line.

File: lib/hsa/mcwamp_hsa.cpp [hunk 2]
```diff
@@ -2203,9 +2276,68 @@ class HSADevice final : public KalmarDevice
     }
 
     void* CreateKernel(const char* fun, Kalmar::KalmarQueue *queue) override {
+        //fun = "_ZZN8hip_impl21grid_launch_hip_impl_IZ76Cijk_Alik_Bljk_SB_MT064x128x08_GSU06_MO10_PGR0_PLR0_TT04_08_WG16_16_01_WGM01PfPKfS3_ffjjjjjjjjjjjjjP12ihipStream_tjPP11ihipEvent_tS8_E4$_36JNS_17Empty_launch_parmERS1_RjSC_SC_SC_SC_SC_EEEvNS_12_GLOBAL__N_119New_grid_launch_tagE4dim3SF_iRKN2hc16accelerator_viewET_DpOT0_ENUlRKNSG_11tiled_indexILi3EEEE_19__cxxamp_trampolineES1_jjjjjj";
```
Comment (whchung) [inline lib/hsa/mcwamp_hsa.cpp]: @bensander please help remove this line.

File: lib/hsa/mcwamp_hsa.cpp [hunk 3]
```diff
@@ -105,6 +106,31 @@ unsigned HCC_DB = 0;
...
+static std::ostream *g_hccProfileStream = nullptr; // point at file or default stream
+
+#define HCC_PROFILE_VERBOSE_BASIC                   (1 << 0)
+#define HCC_PROFILE_VERBOSE_TIMESTAMP               (1 << 1)
+#define HCC_PROFILE_VERBOSE_BARRIER                 (1 << 2)
+
+
+int HCC_PROFILE=0;
+int HCC_PROFILE_VERBOSE=0;
+char * HCC_PROFILE_FILE=nullptr;
```
Comment (whchung) [inline lib/hsa/mcwamp_hsa.cpp:116]: @bensander i'm seeing more file-scope global variables growing... how about we move them into `HSAContext`?

File: lib/hsa/mcwamp_hsa.cpp [hunk 4]
```diff
@@ -395,6 +429,10 @@ class HSACopy : public Kalmar::KalmarAsyncOp {
     hsa_signal_t signal;
     int signalIndex;
     bool isSubmitted;
+    bool isAsync;;          // copy was performed asynchronously
```
Comment (whchung) [inline lib/hsa/mcwamp_hsa.cpp]: @bensander double `;` in the code.

File: lib/hsa/mcwamp_hsa.cpp [hunk 5]
```diff
@@ -4509,8 +4751,6 @@ HSACopy::syncCopyExt(hc::hcCommandKind copyDir, const hc::AmPointerInfo &srcPtrI
             DBOUT(DB_COPY, "HSACopy::syncCopy(), hsa_amd_memory_async_copy() returns: 0x" << std::hex << hsa_status << std::dec <<"\n");
             throw Kalmar::runtime_exception("hsa_amd_memory_async_copy error", hsa_status);
         }
-        Kalmar::ctx.releaseSignal(signal, signalIndex);
```
Comment (whchung) [inline lib/hsa/mcwamp_hsa.cpp:4512]: @bensander could you provide justification why these 2 lines be removed in this PR? I thought this PR is to add kernel profiling logic. but changing these 2 lines would essentially change the behavior of HSA signal reclaim process. I'm not alluding the code is wrong, but perhaps it should be made in a separate PR?

**Review by whchung — CHANGES_REQUESTED** [review]

File: lib/hsa/mcwamp_hsa.cpp (hunk 5 above)
Comment (bensander) [reply lib/hsa/mcwamp_hsa.cpp:4512]: I believe this is not the right place to deallocate a single - instead this should occur in the HSACopy::dispose() routine.  The new profile mode is implemented inside the dispose routine and uses the timestamps from the signal so requires this change.

File: lib/hsa/mcwamp_hsa.cpp (hunk 3 above)
Comment (bensander) [reply lib/hsa/mcwamp_hsa.cpp:116]: Yes - could be separate pull request?  these are environment variables which arguably are best as globals.  the file handle should be moved.

Comment (whchung) [reply lib/hsa/mcwamp_hsa.cpp:116]: ok to make them in another PR.

File: lib/hsa/mcwamp_hsa.cpp (hunk 5 above)
Comment (whchung) [reply lib/hsa/mcwamp_hsa.cpp:4512]: thanks for the clarification

**Review by whchung — APPROVED** [review]

**Pull request merged**